Faultproof withdrawals improvements (#83)
ethereum-optimism/monitorism: op-monitorism/faultproof_withdrawals/runbooks/RUNBOOK.md

## Disclaimer

This runbook may contain references to actions and specifications not included in this repository. This runbook is provided as a guideline for incident response to the scenarios detailed herein, but some details may be redacted or missing due to the sensitive nature of the information. Where information is redacted or missing, we will try to make this clear.

---
## Alerts
An incident will be declared upon receiving an alert. The metrics described below trigger various alerts with differing severities. Each alert necessitates specific actions.

### faultproof-withdrawal-forgery-detected

| **Network** | **Severity Level** | **Impact**                      | **Reaction**                                             | **Actions**                                                                                                                                       |
|-------------|--------------------|---------------------------------|----------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|

> | Mainnet     | SEV1               | Fund loss                       | Immediate action required                                | - Page Faultproof (FP) team<br>- Follow critical incident procedures                                                                               |

Reviewer: Doesn't consistent with Sepolia why in Sepolia we have the 3.5 days and not into the Mainnet?

Author: Removed that extra comment as it was not consistent.

|-------------|--------------------|-----------------------------------------------------|---------------------------------------|--------------------------------------------------------------|
| Mainnet     | SEV2               | - Security max reaction time is reduced<br>- May not be able to react properly to an attack | Daemon is not processing withdrawals  | - Understand the issue with the daemon<br>- If necessary, restart the service |
| Sepolia     | SEV3               | - Security max reaction time is reduced<br>- May not be able to react properly to an attack | Daemon is not processing withdrawals  | - Understand the issue with the daemon<br>- If necessary, restart the service |

##### Alert Description
This alert monitors the number of withdrawal events that are considered normal in a chain. If the number of withdrawal events goes below a specified threshold, we trigger this alert.

#### Investigation
The investigation has two parts:
1. Investigation to triage if the event is a true positive.
2. Investigation to act.

> TODO: Detail the following steps.

Reviewer: Same TODO here.

Author: Removed the sections that are incomplete. Will be added in a second moment

| Mainnet     | SEV2               | - Security max reaction time is reduced<br>- May not be able to react properly to an attack | Too many errors; something may be wrong | - Find out what the error is<br>- Decide if the daemon needs to be patched or configuration changed<br>- If necessary, restart the service |
| Sepolia     | SEV3               | - Security max reaction time is reduced<br>- May not be able to react properly to an attack | Too many errors; something may be wrong | - Find out what the error is<br>- Decide if the daemon needs to be patched or configuration changed<br>- If necessary, restart the service |

##### Alert Description
This alert will be triggered when the number of connection errors goes above a specified threshold. Errors should always be very limited or absent in the monitoring. When present, it often means there is an issue with communication between the monitor and the trusted nodes used for monitoring.

#### Investigation

The investigation has two parts:
1. Investigation to triage if the event is a true positive.
2. Investigation to act.

> TODO: Detail the following steps.

Reviewer: There is still a TODO here, probably need to be fixed.

Author: Removed the sections that are incomplete. Will be added in a second moment

> # Runbook: Incident Response for Faultproof Withdrawals

Reviewer 2: A few questions on how this fits into our broader move towards open sourcing runbooks:
1. Are there internal runbooks in notion that this should replace? If so are those notion runbooks updated accordingly?
2. Is this linked to from the notion table we have that tracks all runbooks?
3. Are all alerts updated to point to this runbook to ensure it's discoverable by oncalls?
4. For open sourced runbooks, did we agree to have them live next to the alert code, as opposed to a single repo that tracks all runbooks?

Author: 1. No this is a new runbook.
2. Not yet, good point, will link it.
3. Alerts are new and points already to this runbook
4. The new agreements is that we share the runbooks via notion first, and in the future in github. This is not mandatory cc: @alfonso-op to confirm.

Reviewer 3: In terms of open-sourcing runbooks, we're going to start with a low-lift trial for now. As @raffaele-oplabs mentioned, starting with open-sourcing a subset of our Notion runbooks, getting feedback from customers / partners and then going from there.

More context [here](https://www.notion.so/oplabs/Operationalizing-New-Protocol-Features-103f153ee16280a1a4f8e8adf8e6a294?pvs=4#118f153ee16280678644c8afecdd7b9f).

Reviewer 2: Got it, great context in there. Thank you both!

> - [Runbook](#runbook)

Reviewer 2: This table of contents looks out of date, I would suggest autogenerating it with doctoc like we do in the monorepo

FaultproofWithdrawal will monitor one L2 chain, let's say op-chain, and one L1 chain, let's say mainnet.
On L2, it will make use of [L2ToL1MessagePasser](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/L2CrossDomainMessenger.sol).
On L1, it will make use of [OptimismPortal2](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L1/OptimismPortal2.sol) and [FaultDisputeGame](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol).

The monitor is driven by the event [WithdrawalProvenExtension1(bytes32 indexed withdrawalHash, address indexed proofSubmitter)](https://github.com/ethereum-optimism/optimism/blob/dd2b21ce786f4c1b722bda270348597182153c8e/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L144C5-L144C102). Every time an event is emitted, the monitor will check if this withdrawal is legitimate or a forgery attempt.

> ## ⚠️ Disclaimer: work in progress!

Reviewer 2: nit, would suggest using markdown callouts for disclaimers like this: https://github.com/orgs/community/discussions/16925

This runbook may contain references to actions and specifications not included in this repository. This runbook is provided as a guideline for incident response to the scenarios detailed herein, but some details may be redacted or missing due to the sensitive nature of the information. Where information is redacted or missing, we will try to make this clear.

The mechanism and the content of the alert is not yet published, so the name of the alerts and the contents of the alerts will depend on your own setup.

## Automated runbooks

> Along side this runbook we have some "automated" runbooks. These runbooks can be used to execute some actions either during triaging of an alert or during containement of an incident. These are basically [Jupiter playbooks](https://jupyter.org/), a mix of executable code and markdown, that makes them perfect for putting together instructions and "executable instructions"

Reviewer 2: nit, these are usually referred to as notebooks
```suggestion
Along side this runbook we have some "automated" runbooks. These runbooks can be used to execute some actions either during triaging of an alert or during containement of an incident. These are basically [Jupiter notebooks](https://jupyter.org/), a mix of executable code and markdown, that makes them perfect for putting together instructions and "executable instructions"
```

> Automated runbooks are in the **automated** subfolder**. 

Reviewer 2: stray asterisks
```suggestion
Automated runbooks are in the **automated** subfolder. 
```

Each runbook is useful for a specific task. After starting the runbook make sure you select the one you need to execute.

---
## Alerts
An incident will be declared upon receiving an alert. The metrics described below trigger various alerts with differing severities. Each alert necessitates specific actions.

### faultproof-withdrawal-forgery-detected

| **Network** | **Severity Level** | **Impact**                      | **Reaction**                                             | **Actions**                                                                                                                                       |
|-------------|--------------------|---------------------------------|----------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|

> | Mainnet     | SEV1               | Potential loss of funds          | Immediate action required                                | - Notify the Faultproof (FP) team<br>- Follow the critical incident procedures (private)                                                                              |

Reviewer 2: Do we / should we have open source definitions of what SEV1 and SEV3 means?

Author: We can do this in a second moment, at the moment the one we use internally are not ready to be opensourced, as they are specific to our case.

> | Sepolia     | SEV3               | Potential loss of funds          | Assess the acceptability of the loss on Sepolia<br>Investigate the specific withdrawal | - Notify the Faultproof (FP) team<br>- Follow the critical incident procedures (⚠️ private procedure) |

Reviewer 2: nit, one row says "(private)" and the other says "(⚠️ private procedure)", we should unify those

2. The [outputRoot provided](https://github.com/ethereum-optimism/optimism/blob/dd2b21ce786f4c1b722bda270348597182153c8e/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L314C15-L314C25) belongs to a FaultDisputeGame that is still [IN_PROGRESS](https://github.com/ethereum-optimism/monitorism/blob/c0b2ecdf4404888e5ceccf6ad14e35c5e5c52664/op-monitorism/faultproof_withdrawals/monitor.go#L336). The metric [faultproof_withdrawals_invalid_proposal_withdrawals_events_count](#faultproof_withdrawals_invalid_proposal_withdrawals_events_count) is used to track such events.

#### Triage Phase

> The alert (⚠️ private alert details) includes the transaction hash that triggered the event. The first step after receiving the alert is to verify whether the attack is real or if it resulted from a monitoring system error or a node issue.

Reviewer 2: What does (⚠️ private alert details) mean here? Is there something this should link to?

Author: In this case we have custom alerts, that are not useful/ready to be opensourced as they are customized to our enviroment and not standalone. In this case here I am just mentioning that the alert itself are not open source.

Reviewer 2: Got it, what do you think of something like this to make that more clear?
```suggestion
The alert (which is currently private) includes the transaction hash that triggered the event. The first step after receiving the alert is to verify whether the attack is real or if it resulted from a monitoring system error or a node issue.
```

To confirm the attack, begin by reviewing the event details and ensuring the conditions for the attack are met.

> You can use the automated `op-monitorism/faultproof_withdrawals/runbooks/automated/triage_potential_attack_event.ipynb` runbook for this process.

Reviewer 2: Do we have instructions on how to run this jupyter notebook somewhere? I suspect not everyone has used them or python before

Author: It is at the beginning of the runbook. Also in the folder there is a readme for how to use with videos.

Reviewer 2: Ah yea that README is perfect. Can we link to that from the runbook?

The jupyter notebook runbook approach is really cool by the way, I like this a lot

### faultproof-potential-withdrawal-forgery-detected

| **Network** | **Severity Level** | **Impact**                      | **Reaction**                                             | **Actions**                                                                                                                                       |
|-------------|--------------------|---------------------------------|----------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| Mainnet     | SEV3               | No immediate impact                       | Investigate the attack, monitor how the attack is proceeding to make sure to be ready to react quickly in case of success                               |  If new type of attack, write a report and eventually create a feature request for improving monitorism monitoring capabilities                                                                            |

> | Sepolia     | SEV3               | No immediate impact                       | Investigate the attack, monitor how the attack is proceeding to make sure to be ready to react quickly in case of success                              | If new type of attack, write a report and eventually create a feature request for improving monitorism monitoring capabilities |

Reviewer 2: Why are they both SEV3 here? It seems surprising for mainnet and testnet to have equal priority

Author: While they do not have immediate impact, and attack on sepolia is equally important to investigate in this case.

##### Alert Description
An event is considered a potential forgery if any of the following conditions apply:
1. The withdrawalHash is not present on L2. We check this by querying [L2ToL1MessagePasser](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/L2CrossDomainMessenger.sol).
2. The [outputRoot provided](https://github.com/ethereum-optimism/optimism/blob/dd2b21ce786f4c1b722bda270348597182153c8e/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L314C15-L314C25) does not match what we see on [L2 block rootState](https://github.com/ethereum-optimism/monitorism/blob/c0b2ecdf4404888e5ceccf6ad14e35c5e5c52664/op-monitorism/faultproof_withdrawals/validator/op_node_helper.go#L47).

and
1. Dispute Game status is IN_PROGRESS

#### Triage Phase

> The alert (⚠️ private alert details) includes the transaction hash that triggered the event. The first step after receiving the alert is to verify whether the attack is real or if it resulted from a monitoring system error or a node issue.

Reviewer 2: same comment on what "(⚠️ private alert details)" is supposed to mean

To confirm the attack, begin by reviewing the event details and ensuring the conditions for the attack are met.

You can use the automated `op-monitorism/faultproof_withdrawals/runbooks/automated/triage_potential_attack_event.ipynb` runbook for this process.

### faultproof-suspicious-withdrawal-forgery-detected

> | **Network** | **Severity Level** | **Impact**                      | **Reaction**                                             | **Actions**                                                                                                                                       |

Reviewer 2: These sections feel repetitive with a lot of the same text. What if we instead consolidated more of this into a single table, where the first column is alert name, another column lists the notebook name, and there's a link to the section in this runbook that contains any alert-specific details

### `faultproof_withdrawals_withdrawals_processed_total`

- **Description:** Number of withdrawals processed. These withdrawals are complete, and they are forgotten.

### `faultproof_withdrawals_initial_l1_height`

- **Description:** Indicates the initial L1 (Layer 1) block height at the start of the monitoring period.

### `faultproof_withdrawals_invalid_proposal_withdrawals_events_count`

- **Description:** Tracks the number of invalid proposal withdrawal events.

### `faultproof_withdrawals_latest_l1_height`

- **Description:** Indicates the latest observed L1 block height.

### `faultproof_withdrawals_next_l1_height`

- **Description:** Represents the next expected L1 block height.

---

## Incident Response

- **Investigate System Logs:** For each alert, review the logs from the affected services to gather more information.
- **Check Node Health:** Verify the status of all nodes involved, especially if connection failures are reported.
- **Coordinate with Relevant Teams:** Communicate with development, operations, and security teams as necessary.
- **Document Findings:** Record the incident details, root cause, and resolution steps for future reference.

---

> ## Conclusion

Reviewer 2: The "Incident Response" and "Conclusion" sections feel out of place. Should "Incident Response" be moved towards the top when explaining how to use the runbook? And "Conclusion" feels like it can be deleted
